src/Compilers/CSharp/Portable/Symbols/Compilation_UsedAssemblies.cs

```diff
@@ -0,0 +1,257 @@
+                                            }
+                                        }
+                                        break;
+
+                                    case RetargetingAssemblySymbol retargetingAssembly:
+                                        usedAssemblies = retargetingAssembly.UnderlyingAssembly.DeclaringCompilation.GetCompleteSetOfUsedAssemblies(cancellationToken);
+                                        if (usedAssemblies is object)
+                                        {
+                                            foreach (AssemblySymbol underlyingDependency in retargetingAssembly.UnderlyingAssembly.SourceModule.ReferencedAssemblySymbols)
+                                            {
+                                                if (!underlyingDependency.IsLinked && usedAssemblies.Contains(underlyingDependency))
+                                                {
+                                                    AssemblySymbol dependency;
+
+                                                    if (!((RetargetingModuleSymbol)retargetingAssembly.Modules[0]).RetargetingDefinitions(underlyingDependency, out dependency))
+                                                    {
+                                                        Debug.Assert(retargetingAssembly.Modules[0].ReferencedAssemblySymbols.Contains(underlyingDependency));
+                                                        dependency = underlyingDependency;
+                                                    }
+
+                                                    addUsedAssembly(dependency, stack);
+                                                }
+                                            }
+                                        }
+
+                                        addReferencedAssemblies(retargetingAssembly, includeMainModule: false, stack);
+                                        break;
+                                    default:
+                                        addReferencedAssemblies(current, includeMainModule: true, stack);
+                                        break;
+                                }
+                            }
+
+                            stack.Free();
+                        }
+                    }
+
+                    if (SourceAssembly.CorLibrary is object)
+                    {
+                        // Add core library
+                        AddUsedAssembly(SourceAssembly.CorLibrary);
+                    }
+                }
+
+                _usedAssemblyReferencesFrozen = true;
+            }
+        }
+
+        internal bool AddUsedAssembly(AssemblySymbol assembly)
+        {
+            if (assembly is null || assembly == SourceAssembly || assembly.IsMissing)
+            {
+                return false;
+            }
+
+            _lazyUsedAssemblyReferences ??= new ConcurrentSet<AssemblySymbol>();
```

Comment:
>_lazyUsedAssemblyReferences [](start = 12, length = 27)

Given that we're locking on this later, I don't think `??=` provides the correct threading guarantee here. We probably need to use `Interlocked.CompareExchange`.